From d8c85abd7826d201a1483d11e6c45e39406c8ea8 Mon Sep 17 00:00:00 2001
From: Lokesh Vutla <lokeshvutla@ti.com>
Date: Wed, 21 Oct 2020 01:19:58 +0530
Subject: [PATCH 06/26] HACK: dma: am65x: Update rchan oes offset with ABI 3.0

Update the rchan oes offset when using ABI 3.0 sysfw.

Note: This patch breaks J721e and that's why its a hack.
To get j721e support cleanly, k3-socinfo driver from 5.4 kernel
should be backported. Since, j721e is not targeted here, not doing
it in a cleaner way and hacking around.

Signed-off-by: Lokesh Vutla <lokeshvutla@ti.com>
---
 drivers/dma/ti/k3-udma.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/ti/k3-udma.c b/drivers/dma/ti/k3-udma.c
index 0f743a5a6a03..c93c0c0f86b5 100644
--- a/drivers/dma/ti/k3-udma.c
+++ b/drivers/dma/ti/k3-udma.c
@@ -1583,6 +1583,7 @@ static int udma_alloc_chan_resources(struct dma_chan *chan)
 	const struct udma_match_data *match_data = uc->ud->match_data;
 	struct udma_tchan *tchan;
 	struct udma_rchan *rchan;
+	u32 rchan_oes_offset;
 	int ret;
 
 	if (uc->pkt_mode || uc->dir == DMA_MEM_TO_MEM) {
@@ -1881,8 +1882,12 @@ static int udma_alloc_chan_resources(struct dma_chan *chan)
 			uc->irq_ra_tisci = k3_ringacc_get_tisci_dev_id(
 								rchan->r_ring);
 			uc->irq_ra_idx = rx_ring;
-			uc->irq_udma_idx = match_data->rchan_oes_offset +
-					   rchan->id;
+			/* ToDo: This will fail for J721E */
+			if (ti_sci_abi_3_and_above(ud->tisci_rm.tisci))
+				rchan_oes_offset = 0x200;
+			else
+				rchan_oes_offset = match_data->rchan_oes_offset;
+			uc->irq_udma_idx = rchan_oes_offset + rchan->id;
 		}
 	}
 
-- 
2.31.1

